design/local-node-agent.md

```diff
@@ -0,0 +1,147 @@
+The agent then uses the Kubernetes API to create a volume attach CRD in the namespace where the Rook cluster providing the storage resides, declaring the intent that an attach needs to happen between a volume and a node.
+   * The attach/detach controller waits on the driver which in turn blocks/waits for the attach to complete on the remote node.
+1. All Rook agents are watching for volume attach CRD events, filtering by a `node=<node name>` label on the CRD instance.
+In response to the creation of the CRD in the previous step, the agent running on the node the attach is for will see the new CRD and perform the attach operation locally on its node.
+Agents running on other nodes will not receive an event since the CRD's `node` label will not match their filter.
+1. After the agent has completed the local attach operation, it will update the CRD with the device path (e.g., `/dev/rbd0`).
+   * The driver on the Controller manager node has been blocking/waiting for this update and will now finally return from the `Attach()` call.
+1. `Mount()` is called in the Rook Flexvolume driver by the Kubelet on the node that will be consuming the storage.
+The driver directly performs the mount operation with no need to call into the Rook agent.
+If necessary, the driver will also format the volume according to the filesystem type expressed on the storageclass for the volume.
+   * This operation is synchronous and `Mount()` will not return until the mount/format is completed.
+
+#### Usage of CRDs
+The control flow described above explained how during the `Attach()` phase, the Rook agent (running on the Controller manager node) makes use of a volume attach CRD in order to convey to the agent running on the node that will consume the storage that there is a need to attach a volume to that node.
+This usage of the Kubernetes API is a major reason why the Flexvolume driver, which does not run in a context that has any cluster credentials, is insufficient for this design, and thus why a Rook agent pod is desirable.
+Let's enumerate some of the benefits of using a volume attach CRD:
+
+1. The admin gets a native `kubectl` experience for viewing and getting information on volume attachments.
+1. The CRD helps provide fencing for the volume in a generalized way, it is not specific to the underlying storage Rook is creating and managing.
+The existence of the CRD provides a means of bookkeeping to signal that the volume is locked and in use.
+1. A CRD removes any need for an API surface area on the agent.
+When the Flexvolume driver and Rook agent on the Controller manager node need to notify the agent on another node to perform the attach, a CRD provides a native Kubernetes means of communication, instead of the agent having to expose and secure a REST endpoint.
+
```

Comment:
I would add that since we dont have the need for a restful API, we can leverage the K8S authorization model RBAC.

Reply:
Added a note about using RBAC for free.